### ISSUE TITLE
Going from keras model to dlib model

### ISSUE THREAD

**Issue by Jendker** [davisking/dlib#1593]
I know that this is not strictly dlib related question and maybe I will not get the answer, but still wanted to try, the audience is pretty big here :)

I am trying to migrate from keras to dlib, because of the superior documentation and ease of metric training here. Right now I am trying to build the same network and get comparable results of the same model, but the dlib model does not get trained properly. The average loss goes from 4.63 to 4.55 on the first epoch and afterwards it does not get any further. I even tried with increasing `trainer.set_iterations_without_progress_threshold(10000)` but it still does not get the loss lower than 4.53... Am I missing something here?

I am feeding the data in dlib as a `std::vector<matrix<double, 0, 1>>` where `matrix` comprises of 340 row vector.

The network:

    using net_type = loss_multiclass_log<
            fc<286,
            dropout<relu<fc<1500,
            dropout<relu<fc<1500,
            input<matrix<double, 0, 1>>>>>>>>>>;

### The dlib output:

layer<0>	loss_multiclass_log
layer<1>	fc	 (num_outputs=286) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0
layer<2>	dropout	 (drop_rate=0.275)
layer<3>	relu
layer<4>	fc	 (num_outputs=1500) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0
layer<5>	dropout	 (drop_rate=0.275)
layer<6>	relu
layer<7>	fc	 (num_outputs=1500) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0
layer<8>	input<matrix>
Epoch: 1     learning rate: 0.01  average loss: 4.6328       steps without apparent progress: 2410
Epoch: 2     learning rate: 0.01  average loss: 4.55244      steps without apparent progress: 5213
Epoch: 3     learning rate: 0.01  average loss: 4.5501       steps without apparent progress: 7901
Epoch: 4     learning rate: 0.01  average loss: 4.54932      steps without apparent progress: 9152
Epoch: 5     learning rate: 0.001  average loss: 4.54276      steps without apparent progress: 5283
Epoch: 6     learning rate: 0.001  average loss: 4.53794      steps without apparent progress: 6098
Epoch: 7     learning rate: 0.001  average loss: 4.53792      steps without apparent progress: 8740
Epoch: 8     learning rate: 0.0001  average loss: 4.53726      steps without apparent progress: 9715
Epoch: 9     learning rate: 1e-05  average loss: 4.5363       steps without apparent progress: 9730
Epoch: 10    learning rate: 1e-06  average loss: 4.53364      steps without apparent progress: 0
training num_right: 5835
training num_wrong: 91605
training accuracy:  0.059883
testing num_right: 1460
testing num_wrong: 22900
testing accuracy:  0.0599343

### Keras model and results:
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
_________________________________________________________________
dense_1 (Dense)              (None, 1500)              505500    
_________________________________________________________________
activation_1 (Activation)    (None, 1500)              0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 1500)              0         
_________________________________________________________________
dense_2 (Dense)              (None, 1500)              2251500   
_________________________________________________________________
activation_2 (Activation)    (None, 1500)              0         
_________________________________________________________________
dropout_2 (Dropout)          (None, 1500)              0         
_________________________________________________________________
dense_3 (Dense)              (None, 286)               429286    
_________________________________________________________________

epoch,acc,loss
0,0.6818555008210181,1.3333302831659568
1,0.8814347290640394,0.42806850175919203
2,0.9182881773399014,0.2863174603056095
3,0.9363505747126437,0.2230666387248142
4,0.9461720032840722,0.18629631130621205
5,0.952524630541872,0.1625511203678432
6,0.9577380952380953,0.14498288792539663
7,0.9617918719211823,0.12999034807233564
8,0.9647167487684729,0.12031887783768337
9,0.9670566502463054,0.11327744584053843

### If there is anything more I should provide I will happily do so.

**Comment by davisking** [comment]
It’s hard to tell based on what you posted. Maybe the learning rate should be higher. Maybe you set the weight decay to something way too high. Maybe you just aren’t loading the labels and it’s training on random labels.

**Comment by Jendker** [comment]
`weight_decay_mult` is at 1 so I guess this is not a problem. I will try with the higher learning rate and confirm how it behaves, thank you!

**Comment by Jendker** [comment]
I am not sure what was it, I assume some numerical instability, because when I removed some data columns in the samples to remove columns with very low variance and applied the distribution normalization the network is being trained properly. So success :)

But I had some problems with the distribution normalization:
here I wanted to apply the normalization as in example 'rank_features_ex.cpp':
```
 const sample_type m(mean(mat(samples)));  // compute a mean vector
 const sample_type sd(reciprocal(stddev(mat(samples)))); // compute a standard deviation vector
 // now normalize each sample
 for (unsigned long i = 0; i < samples.size(); ++i)
     samples[i] = pointwise_multiply(samples[i] - m, sd); 
```
but `sd` vector had some values which were difficult to interpret, some `inf`s and 0s where the shouldn't be.

So I applied it in a different way:

```
   const auto m(dlib::mean(dlib::mat(dlib_data)));  // compute a mean vector
   const auto sd(dlib::stddev(dlib::mat(dlib_data))); // compute a standard deviation vector
   // manually do reciprocal
   std::vector<double> sd_vector(sd.begin(), sd.end());
   std::for_each(sd_vector.begin(), sd_vector.end(), [](double& v){v = 1.0 / v;});
   const auto sd_reciprocal(dlib::mat(sd_vector));
    // now normalize each sample
    for (unsigned long i = 0; i < samples.size(); ++i)
        samples[i] = pointwise_multiply(samples[i] - m, sd_reciprocal); 
```
by manually making reciprocal of each element. That was it what made the network train properly.
Why did this `dlib::reciprocal` not work properly? Is it the problem, that the data has size NC = 1 and NR = 0? I just create data structure as it was in 'dnn_metric_learning_ex.cpp': `std::vector<matrix<double,0,1>> samples;`

**Comment by davisking** [comment]
`reciprocal` does just what it sounds like it does, you must have a bug in the code.  Please post a complete and minimal program that reproduces the error if you think you found a bug in `reciprocal`.

**Comment by Jendker** [comment]
Perfect, I'll try to prepare an example and post it in a separate issue, which description will better suit the case.
